### ISSUE TITLE
Destroy callback passes false to Comparable <=> method

### ISSUE THREAD

**Issue by jhaungs** [rails/rails#24584]
### Steps to reproduce

```
begin
  require 'bundler/inline'
rescue LoadError => e
  $stderr.puts 'Bundler version 1.10 or later is required. Please update your Bundler'
  raise e
end

gemfile(true) do
  source 'https://rubygems.org'
  gem 'rails', github: 'rails/rails'
  gem 'sqlite3'
end

require 'active_record'
require 'minitest/autorun'
require 'logger'

# This connection will do for database-independent bug reports.
ActiveRecord::Base.establish_connection(adapter: 'sqlite3', database: ':memory:')
ActiveRecord::Base.logger = Logger.new(STDOUT)

ActiveRecord::Schema.define do
  create_table :posts, force: true  do |t|
    t.string :name
  end

  create_table :comments, force: true  do |t|
    t.integer :post_id
  end
end

class Post < ActiveRecord::Base
  include Comparable
  after_destroy :log_delete
  has_many :comments
  
  def <=>(other)
    # return 0 unless other  # This will mask the problem.
    self.name <=> other.name
  end
  
  def log_delete
    puts "Logging #{self.class} #{attributes.inspect}"
    true
  end
end

class Comment < ActiveRecord::Base
  belongs_to :post
end

class BugTest < Minitest::Test
  def test_association_stuff
    post = Post.create!
    post.comments << Comment.create!

    assert_equal 1, post.comments.count
    assert_equal 1, Comment.count
    assert_equal post.id, Comment.first.post.id
    
    post.destroy   # This will crash, passing false to `other` in <=>
  end
end

```

### Expected behavior
Comparison method parameter should be an ActiveRecord object of the same class.

### Actual behavior
Callbacks.rb sends false to the <=> comparison method

### System configuration
**Rails version**: master and 4.2.6

**Ruby version**: 2.3.0

**Comment by maclover7** [comment]
It looks like ActiveRecord::Core [also defines a `<=>` method](https://github.com/rails/rails/blob/master/activerecord/lib/active_record/core.rb#L455) -- not sure if it's advised to overwrite internal methods like that :grimacing:

thoughts @sgrif?

**Comment by sgrif** [comment]
> not sure if it's advised to overwrite internal methods like that

Overwriting it is fine, but you need to adhere to the same contract. In this case it looks like we're expected to get objects of other classes, so your implementation should handle the same.

**Comment by jhaungs** [comment]
I would still maintain that this is a bug in the callback comparison to **false.**  Perhaps it should be using **===** instead, as it seems to be looking for a sentinel value, not an A/R object.

This usage of overriding <=> came from Rails 1 or 2, and pre-dates the addition of the <=> method in 2013. In my case, the code is eight years old. 

A more recent example (Rails 4.2.6 documentation): 

http://api.rubyonrails.org/classes/ActiveRecord/Aggregations/ClassMethods.html

The suggested usage of <=> on this page would also crash if a callback passed **false** as _other_money_ into the example <=> method:

```

def <=>(other_money)
    if currency == other_money.currency
      amount <=> other_money.amount
    else
      amount <=> other_money.exchange_to(currency).amount
    end
  end
```

**Comment by jeremy** [comment]
@jhaungs That example is buggy for the same reason.

[Comparable should return `nil` if the other object is not comparable](http://ruby-doc.org/core-2.3.0/Comparable.html):
> The Comparable mixin is used by classes whose objects may be ordered. The class must define the <=> operator, which compares the receiver against another object, returning -1, 0, or +1 depending on whether the receiver is less than, equal to, or greater than the other object. If the other object is not comparable then the <=> operator should return nil.

**Comment by jeremy** [comment]
That said, it's easy to implement this behavior incompletely in apps. From their point of view, a stable Rails upgrade breaks their apps when it's really just awakened a long-dormant bug.

It could be worth changing our implementation to avoid tickling app bugs since it's clear that this is a fairly widespread error.

**Comment by tobymurray-nanometrics** [comment]
This is long since closed, but I just ran into this as well in 4.2.7.1. Neither the documentation for Ruby nor [Rails](http://api.rubyonrails.org/classes/ActiveRecord/Aggregations/ClassMethods.html) (which was linked above, and remains incorrect according to this discussion) lends itself to implementing this properly. The [current RubyDoc](https://ruby-doc.org/core-2.2.3/Comparable.html) has the following implementation (which is exactly how this bug shows up):

```
def <=>(anOther)
  str.size <=> anOther.str.size
end
```

If anyone else runs into this, the (now slightly out of date) link above yields the following implementation in ActiveRecord:

```
def <=>(other_object)
  if other_object.is_a?(self.class)
    self.to_key <=> other_object.to_key
  else
    super
  end
end
```

So essentially ActiveRecord has co-opted `<=>`, and you should either avoid using it or use the same format as ActiveRecord. Not something particularly intuitive to track down...
